astropy/nddata/decorators.py

```diff
@@ -3,161 +3,236 @@
+            # that can be passed as kwargs.
             if unpack:
-
-                ignored = []
-
                 # We loop over a list of pre-defined properties
-                for prop in supported_properties:
-
-                    # We only need to do something if the property exists on the
-                    # NDData object
-                    if hasattr(data, prop):
+                for prop in islice(SUPPORTED_PROPERTIES, 1, None):
+                    # We only need to do something if the property exists on
+                    # the NDData object
+                    try:
                         value = getattr(data, prop)
-                        if (prop == 'meta' and len(value) > 0) or (prop != 'meta' and value is not None):
-                            if prop in func_kwargs:
-                                if prop in kwargs and kwargs[prop] is not None:
-                                    warnings.warn("Property {0} has been passed explicitly and as an "
-                                                  "NDData property, using explicitly specified value".format(prop),
-                                                  AstropyUserWarning)
-                                else:
-                                    kwargs[prop] = value
-                            else:
-                                ignored.append(prop)
+                    except AttributeError:
+                        continue
+                    # Skip if the property exists but is None or empty.
+                    if prop == 'meta' and not value:
+                        continue
+                    elif value is None:
+                        continue
+                    # Warn if the property is set but not used by the function.
+                    propmatch = attr_arg_map.get(prop, prop)
+                    if propmatch not in func_kwargs:
+                        ignored.append(prop)
+                        continue
+
+                    # Check if the property was explicitly given and issue a
+                    # Warning if it is.
+                    if propmatch in kwargs:
+                        # If it's in the func_args it's trivial but if it was
+                        # in the func_kwargs we need to compare it to the
+                        # default.
+                        # FIXME: This obviously fails if the explicitly given
+                        # value is identical to the default. No idea how that
+                        # could be achieved without replacing the signature or
+                        # parts of it. That would have other drawbacks like
+                        # making double-decorating nearly impossible and
+                        # violating the sphinx-documentation for that function.
+                        if (propmatch in func_args or
+                                (propmatch in func_kwargs and
+                                 not np.array_equal(kwargs[propmatch],
```

Comment:
It works with `WCS` because it does an element-wise comparison of object-arrays then. But I've noticed some classes (`fits.Header`) when it doesn't work. I do have an idea how to make it work though. Will push a fix later.